### ISSUE TITLE
lsp--diagnostics should be per workspace (was: shouldn't the lsp--diagnostics variable be defvar-local?)

### ISSUE THREAD

**Issue by Havner** [emacs-lsp/lsp-mode#390]
As in title. I've found this variable to gain weight pretty significantly on large projects (measured even in GBs). Closing the projects doesn't reclaim this memory. And its description suggests it's per file.
I've marked it local for the test and it seems to work fine.
I'm not sure that is correct, but after I switched to using LSP (with cquery) I've found my Emacs to consume lots of memory at times. And the culprit is this particular variable.

**Comment by yyoncho** [comment]
`lsp--diagnostigs` is used by `lsp-ui-flycheck` when rendering errors it should not be buffer local. But there is an issue indeed - `lsp--diagnostigs` must be part of `lsp--client` or `lsp--workspace` structures and it should have the same lifetime as the workspace.

**Comment by Havner** [comment]
Yeah, I know it's used for that but I still don't get why it can't be buffer local. Flycheck errors are per file/buffer.

**Comment by yyoncho** [comment]
It stores the errors for all of the files in the workspace. Actually in any workspace.

**Comment by Havner** [comment]
And why does flycheck need errors from other buffers to display them? :-)

**Comment by yyoncho** [comment]
Sorry for being unclear - the errors are reported for all of the buffers by LSP server. So when you switch to a buffer it displays the errors from that buffer(the already reported one). As a side note, flycheck is capable of displaying errors from multiple buffers.

**Comment by Havner** [comment]
Ok, got it, so for the first point it goes asynchronously. Thanks for explanation.
Anyway, the issue for high memory usage still remains.
This also is especially prominent for big repos where there is no compile_commands.json and presumably there is lots of errors due to missing headers.
I've opened kernel source once this way and emacs memory usage went to 6GB before I had to kill it. And it was still growing. Making this variable per workspace won't fix that.

**Comment by D4N** [comment]
> Anyway, the issue for high memory usage still remains.

I can confirm this, as I have recently tried opening the Kernel version 2.6.32 sources (which wouldn't compile with gcc 8 and produced a bazillion of warnings). Emacs with lsp-mode was consuming about 5-10 times more memory then cquery was, before I had to kill emacs. I was only able to use cquery on the kernel with eglot (but uses flymake instead of flycheck).

**Comment by vibhavp** [comment]
Making it buffer local now would probably break API compatibility with packages using previous package version. Introducing a new variable `lsp--buffer-diagnostics` would be a better idea, IMO.

> Emacs with lsp-mode was consuming about 5-10 times more memory then cquery was, before I had to kill emacs

I think this was because `lsp--diagnostics` is storing diagnostics for *all files *in the kernel source code, which obviously would result in a large amount of memory being used. A partial fix to this would be to only store diagnostics for files that are currently open.

**Comment by D4N** [comment]
Vibhav Pant <notifications@github.com> writes:

> Making it buffer local now would probably break API compatibility with packages using previous package version. Introducing a new variable `lsp--buffer-diagnostics` would be a better idea, IMO.
>
>> Emacs with lsp-mode was consuming about 5-10 times more memory then cquery was, before I had to kill emacs
>
> I think this was because `lsp--diagnostics` is storing diagnostics for
> *all files *in the kernel source code, which obviously would result in
> a large amount of memory being used. A partial fix to this would be to
> only store diagnostics for files that are currently open.

That sounds like it could fix the issue, as I highly doubt that anyone
would want to open all *files* from any larger project.

**Comment by vibhavp** [comment]
@D4N I just pushed d17c55448c26c8475a5e38354ca42d7875b239b4, could you test it and see if it solves your memory issues?

**Comment by yyoncho** [comment]
@vibhavp  d17c554 breaks the option to check what are the errors/warnings on project level(e. g. using `lsp-ui-flycheck-list`). Also, if the buffer is closed on flycheck reporting you wont get the errors when you open it since they are already published. I propose to make this behaviour optional.

**Comment by MaskRay** [comment]
@vibhavp 
The backward compatibility for diagnostics should be very little of concern. What people complain (from various sources I am aware of) is why sometimes flycheck diagnostics linger around after the code has been fixed (probably some `(flycheck-clear-errors)` issue), diagnostics appear in unrelated files, etc. Some breaking change is unavoidable but I don't think that would make users more painful. I've checked lsp-haskell lsp-java emacs-ccls emacs-cquery I think only lsp-ui uses the diagnostics API.
